src/catchup/ApplyBucketsWork.cpp

```diff
@@ -192,8 +232,15 @@ ApplyBucketsWork::onRun()
         return State::WORK_RUNNING;
     }
 
-    CLOG(DEBUG, "History") << "ApplyBuckets : done, restarting merges";
+    CLOG(INFO, "History") << "ApplyBuckets : done, restarting merges";
     mApp.getBucketManager().assumeState(mApplyState, mMaxProtocolVersion);
+
+    if (mResolveMerges)
```

Comment:
This logic feels like it's in a strange order to me: the _first_ block checks for `mResolveMerges && mDelayTimer` and if so handles the resolving merge, then the _second_ block checks `mResolveMerges` and if so sets up `mDelayTimer`. This feels like the blocks are kinda backwards and mis-nested, perhaps? Would logic like this work? I think it's the same, but reads top-to-bottom in the order of return-visits-to-the-function.

```
if (mResolveMerges) {
    if (!mDelayTimer) {
        // This is our first time here, start the work
        bm.assumeState(...);
        mDelayTimer = make_unique<VirtualTimer>();
        return WORK_RUNNING;
    } else {
       // We're returning to running work, check to see if it's done
       bl.resolveAnyReadyFutures();
       if (!bl.futuresAllResolved()) {
           // Not yet done, queue a wait and callback
           mDelayTimer = ...
           mDelayTimer->expires_from_now(...)
           return WORK_WAITING;
        }
    }
}
// all done
return WORK_SUCCESS;

```

Reply:
Update: I misunderstood that there's a big block of code between timer start and continue that github elided in the review view, this can't be readily reordered.